src/spawn.rs

```diff
@@ -0,0 +1,248 @@
+    let (rc, id) = Spawn::default().spawn_cell(ckb_sim, argc, argv, inherited_fds);
+    if let Some(id) = id {
+        unsafe { *({ pid }) = id.into() };
+    }
+
+    rc
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_wait(_pid: u64, _code: *mut i8) -> c_int {
+    panic!("unsupport");
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_process_id() -> u64 {
+    Spawn::default().pid()
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_pipe(fds: *mut u64) -> c_int {
+    Spawn::default().pipe(fds)
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_read(fd: u64, buf: *mut c_void, length: *mut usize) -> c_int {
+    Spawn::default().read(fd.into(), buf, length)
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_write(fd: u64, buf: *const c_void, length: *mut usize) -> c_int {
+    Spawn::default().write(fd.into(), buf, length)
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_inherited_fds(fds: *mut u64, length: *mut usize) -> c_int {
+    Spawn::default().inherited_fds(fds, length)
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_close(fd: u64) -> c_int {
+    Spawn::default().close(fd)
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_load_block_extension(
+    _addr: *mut c_void,
+    _len: *mut u64,
+    _offset: usize,
+    _index: usize,
+    _source: usize,
+) -> c_int {
+    panic!("unsupport");
+}
+
+struct Spawn {
+    tx_id: TxID,
+    proc_id: ProcID,
+}
+
+const MAX_FDS: usize = 64;
+
+impl Default for Spawn {
+    fn default() -> Self {
+        Self {
+            tx_id: TxContext::ctx_id(),
+            proc_id: Process::ctx_id(),
+        }
+    }
+}
+impl Spawn {
+    fn pid(&self) -> u64 {
+        self.proc_id.clone().into()
+    }
+    fn pipe(&self, fds: *mut u64) -> i32 {
+        if get_tx!(&self.tx_id).len_pipe() >= MAX_FDS {
+            return 9; // MAX_FDS_CREATED
+        }
+
+        let out = get_tx_mut!(&TxContext::ctx_id()).new_pipe();
+        Self::copy_fd(&[out.0, out.1], fds);
+        0
+    }
+    fn close(&self, fd: u64) -> i32 {
+        let fd = fd.into();
+
+        let r = get_tx_mut!(self.tx_id).close_pipe(fd);
+        if r {
+            0
+        } else {
+            6 // CKB_INVALID_FD
+        }
+    }
+
+    fn spawn_cell(
+        &self,
+        ckb_sim: utils::CkbNativeSimulator,
+        argc: i32,
+        argv: *const *const u8,
+        inherited_fds: *const u64,
+    ) -> (i32, Option<ProcID>) {
+        let new_id = self.new_process_id(inherited_fds);
+        ckb_sim.ckb_std_main_async(argc, argv, &new_id);
+
+        let event = crate::get_cur_proc!().wait_by_pid(&new_id);
+        event.wait();
+
+        (0, Some(new_id))
+    }
+    fn inherited_fds(&self, fds: *mut u64, length: *mut usize) -> i32 {
+        let out_fds = get_tx!(&TxContext::ctx_id())
+            .process(&Process::ctx_id())
+            .inherited_fds();
+        let len = out_fds.len().min(unsafe { *length });
+
+        Self::copy_fd(&out_fds[0..len], fds);
+        0
+    }
+
+    fn read(&self, fd: Fd, buf: *mut c_void, length: *mut usize) -> c_int {
```

Comment:
If 100 bytes of data are written from one end of the pipe and the other end performs the following actions:
1. Reads 30 bytes
2. Reads 40 bytes
3. Reads 30 bytes

The simulator is unable to handle this correctly.